### PR TITLE
H195: per-channel best-of-K EMA selection across full cohort

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

For every completed cohort run, **different test channels may have best test accuracy at different training EPs** — and the val_abupt-best EP (terminal or EP9-by-val) likely doesn't coincide with the test-best EP for most channels. This systematic val-test EP mismatch is a direct manifestation of the slope-flattening pathology (Findings E, F, G from today's H183/H185 closures).

H195 maps the **per-channel val-best EP vs test-best EP gap** across all 9 completed cohort runs and answers:
1. Is the EP gap systematic (always negative: test-best EP < val-best EP)?
2. Is the gap channel-specific (WSS_x might peak at EP8 while SP peaks at EP13)?
3. Does any (run, EP) combination meet all 4 merge gates + WSS_x basin intact?

## Instructions

**This is a pure eval sprint — NO training. No source file modifications.**

### Step 1: Enumerate cohort runs

| Run | W&B run ID | Status |
|---|---|---|
| H112 | `u9ue2ryb` | Completed, terminal logged |
| H164e | `zrv3dasr` | Completed, terminal logged |
| H183 | `5k58uzqc` | Completed, terminal logged |
| H184 | `i2rl27nj` | Completed, terminal logged |
| H185 | (get from PR #1353 SENPAI-RESULT) | Completed, terminal logged |
| H188 | `18t5rx2t` | Completed, terminal logged |
| H189 | `bp12b1cc` | Completed, terminal logged |
| H190 | `9f2jtrg2` | May still be training — check status |
| H191 | `5y5a5tgr` | Completed, terminal logged |

### Step 2: W&B val screening (no GPU needed)

For each run, pull val_abupt from W&B history at each saved EMA checkpoint (EP6, EP7, EP8, EP9, EP10, EP11, EP12, EP13). Build a 9×8 table of val_abupt × EP.

Filter to checkpoints where val_abupt < 6.20% (generous slack). These are Stage 2 candidates.

### Step 3: Test eval on Stage 2 candidates (DDP=8 GPUs)

For each candidate (likely ~15-25 total across all runs):

```bash
python -m torch.distributed.run --nproc-per-node=8 \
  train.py \
  --eval_only \
  --checkpoint <path-to-EMA-ckpt> \
  --wandb_group h195-tanjiro-perchannel-bofk \
  --run_name h195-<run>-ep<N>
```

Collect per-channel test metrics: test_WSS, test_WSS_x, test_WSS_y, test_WSS_z, test_VP, test_SP.

### Step 4: Per-run analysis

For each run, find:
- val-best EP (by val_abupt)
- test-best EP per channel (test_WSS, test_WSS_x, test_WSS_y, test_WSS_z, test_VP, test_SP)
- EP gap = val-best EP minus test-best EP (negative means test-best is EARLIER in training)

### Step 5: Aggregate and report

1. Table: 9 runs × (val-best EP, test-best EP, EP gap, WSS_x slope at test-best EP)
2. Is EP gap consistently negative? (If yes: we should always use a mid-EP checkpoint, not terminal)
3. Is it channel-specific? (Different EP peaks per channel)
4. SOTA candidate: any (run, EP) with test_WSS ≤ 6.752 AND test_VP ≤ 3.421 AND test_SP ≤ 3.577 AND WSS_x slope ≤ −0.020pp

## Coordination

- Askeladd (H198) is running global test_WSS minimization across the cohort — overlapping evals may be reused
- Alphonse (H204) covers H183 pre-flip specifically; frieren (H205) covers H185 pre-flip
- H195 is BROADER: all 9 runs × per-channel EP gap analysis

If askeladd's H198 produces test evals for same candidates, reuse those W&B logs instead of re-running.

## Baseline

Current best (H112, PR #1283, W&B run `u9ue2ryb`):
- val_abupt: **6.1358%**
- test_WSS: **6.752%**
- test_VP: **3.421%**
- test_SP: **3.695%** (strict gate ≤ 3.577%)

All gates for SOTA candidate:
- val_abupt < 6.1358%
- test_WSS ≤ 6.752%
- test_VP ≤ 3.421%
- test_SP ≤ 3.577%
- WSS_x slope ≤ −0.020pp

Expected runtime: ~1.5-2h on 8 GPUs. Stage 1 (W&B query) ~10min. Stage 2 (~15-25 evals × ~25min) is the bottleneck, but candidates are spread across 9 runs so some fast exits occur.

If WINNER found: tag advisor for merge-winner skill.

## Scientific value

Even with no merge: this delivers a **systematic per-channel EP transfer characterization** across the full cohort — directly answers "is the terminal checkpoint always wrong, or is it run-specific?" This would be cited in any paper writeup of this program.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1377 because ADVISOR — H195 closure decision: **CLOSE — SAME INFRASTRUCTURE BLOCKER AS H204; PIVOT**.

## Decision: CLOSE before GPU waste

Alphonse just discovered (PR #1372) that **mid-EP EMA checkpoints were never saved across the cohort**. W&B artifact aliases for every completed run are `['best', 'latest', 'epoch-13']` only.

This BLOCKS H195's per-channel best-of-K EMA selection within each run (each run has K=1 EMA snapshot).

## Pivot: H210 — Cross-recipe SWA on ALL 5 best EP13 artifacts

Using only the EP13 best artifacts from completed runs (H112, H164e, H183, H185, H190 — five distinct recipe replicates), **average the state_dicts into a single set of weights**. This is intra-program cross-recipe weight averaging — different from frieren's H193 (H112+H164e only).

```
W_swa = (W_H112 + W_H164e + W_H183 + W_H185 + W_H190) / 5
```

Hypothesis: cross-recipe consensus across 5 diverse recipes finds a flatter minimum than any single recipe, potentially resisting the val→test slope-flattening pathology that affected every individual cohort run.

Bonus arms: try 3-subset variants (H112 + H164e + H183) and (H112 + H185 + H190) — narrows attribution of which recipes contribute. Pure eval, ~1-1.5h on 8 GPUs.

**See new PR for full H210 instructions.** Closing this PR.

ADVISOR.
